File: versions/v10/spec_format/spec_decoder.py
"""
Spectrum Algo — .spec Decoder v1
Converts a .spec binary file back into source code.

See spec_encoder.py for the full format specification.
"""

import sys
import struct
import zlib
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
import dictionary as D
from spec_format.extension_tokens import extension_id_to_literal
from tokenizers.text_tokenizer import reconstruct_text
from spec_format._frozen import (
    get_id_to_token_for_version,
    get_ascii_base_for_version,
    MIN_SUPPORTED_VERSION,
)

LANGUAGE_TEXT = 4
LANGUAGE_XML = 9

# ─────────────────────────────────────────────────────────────────────────────
# Constants (must match encoder)
# ─────────────────────────────────────────────────────────────────────────────
MAGIC    = b'SPEC'
FLAG_RLE = 0b0000_0001
HEADER_SIZE = 16


class SpecFormatError(Exception):
    pass


# ─────────────────────────────────────────────────────────────────────────────
# Header parsing
# ─────────────────────────────────────────────────────────────────────────────

def parse_header(data: bytes) -> dict:
    if len(data) < HEADER_SIZE:
        raise SpecFormatError("File too short to contain a valid header.")
    if data[:4] != MAGIC:
        raise SpecFormatError(
            f"Bad magic bytes: expected {MAGIC!r}, got {data[:4]!r}. "
            "Is this a .spec file?")

    dict_version,  = struct.unpack_from(">H", data, 4)
    flags,         = struct.unpack_from(">H", data, 6)
    orig_length,   = struct.unpack_from(">I", data, 8)
    language_id,   = struct.unpack_from(">H", data, 12)
    checksum,      = struct.unpack_from(">H", data, 14)

    return {
        "dict_version": dict_version,
        "flags":        flags,
        "orig_length":  orig_length,
        "language_id":  language_id,
        "checksum":     checksum,
        "rle_enabled":  bool(flags & FLAG_RLE),
    }


# ─────────────────────────────────────────────────────────────────────────────
# ID stream → tokens
# ─────────────────────────────────────────────────────────────────────────────

def ids_to_tokens(
    ids: list[int],
    id_to_token: dict[int, str] | None = None,
    ascii_base: int | None = None,
) -> list[str]:
    """
    Decode a uint32 ID stream (with optional RLE markers) back to token strings.

    id_to_token  — version-specific ID→token map; defaults to current dict.
    ascii_base   — version-specific ASCII fallback base ID; defaults to current.

    ID scheme:
      0 … N-1                  → dictionary token
      N … N+127                → ASCII char (ord = ID - N)
      SPEC_ID_RLE  (0xFFFFFFFD) → repeat previous token (next ID = count)
      SPEC_ID_UNICODE (0xFFFFFFFE) → Unicode char > 127 (next ID = code point)
    """
    if id_to_token is None:
        id_to_token = D.SPEC_ID_TO_TOKEN
    if ascii_base is None:
        ascii_base = D.SPEC_ID_ASCII_BASE

    tokens: list[str] = []
    last_tok: str | None = None
    i = 0
    n = len(ids)

    while i < n:
        val = ids[i]

        # RLE marker
        if val == D.SPEC_ID_RLE:
            if i + 1 >= n:
                raise SpecFormatError(f"RLE marker at position {i} has no count.")
            count = ids[i + 1]
            if last_tok is not None:
                tokens.extend([last_tok] * count)
            i += 2
            continue

        # Unicode fallback (> 127)
        if val == D.SPEC_ID_UNICODE:
            if i + 1 >= n:
                raise SpecFormatError(f"Unicode escape at {i} is truncated.")
            cp = ids[i + 1]
            tok = chr(cp)
            tokens.append(tok)
            last_tok = tok
            i += 2
            continue

        # ASCII fallback (IDs ascii_base … ascii_base+127)
        if ascii_base <= val < ascii_base + 128:
            tok = chr(val - ascii_base)
            tokens.append(tok)
            last_tok = tok
            i += 1
            continue

        # Extension-library token ranges
        extension_literal = extension_id_to_literal(val)
        if extension_literal is not None:
            tokens.append(extension_literal)
            last_tok = extension_literal
            i += 1
            continue

        # Dictionary token
        tok = id_to_token.get(val)
        if tok is None:
            raise SpecFormatError(
                f"Unknown token ID {val} at stream position {i}. "
                f"The file may have been encoded with a newer dictionary version "
                f"than available. Try upgrading Spectrum Algo.")
        tokens.append(tok)
        last_tok = tok
        i += 1

    return tokens


# ─────────────────────────────────────────────────────────────────────────────
# Top-level decode
# ─────────────────────────────────────────────────────────────────────────────

def decode_file(spec_path: str, output_path: str) -> dict:
    """
    Decode a .spec file back to source code.

    Returns a stats dict including fidelity information.
    """
    spec_path   = Path(spec_path)
    output_path = Path(output_path)

    raw = spec_path.read_bytes()
    print(f"[spec_dec] Loaded {spec_path.name}  ({len(raw):,} bytes)")

    # Parse header
    meta = parse_header(raw)
    dict_version = meta["dict_version"]
    orig_length  = meta["orig_length"]
    expected_cksum = meta["checksum"]

    print(f"[spec_dec] Header: dict_v={dict_version}, "
          f"orig={orig_length:,} B, "
          f"rle={meta['rle_enabled']}, "
          f"lang={meta['language_id']}")

    # ── Version compatibility check ──────────────────────────────────────────
    if dict_version < MIN_SUPPORTED_VERSION:
        raise SpecFormatError(
            f"Dict version {dict_version} is below the minimum supported "
            f"version ({MIN_SUPPORTED_VERSION}). Re-encode from original source."
        )

    if dict_version == D.DICT_VERSION:
        # Current version — use live dict directly (fastest path)
        id_to_token = D.SPEC_ID_TO_TOKEN
        ascii_base  = D.SPEC_ID_ASCII_BASE
        compat_note = ""
    elif dict_version < D.DICT_VERSION:
        # Older file — load the frozen snapshot for that version
        id_to_token = get_id_to_token_for_version(dict_version)
        ascii_base  = get_ascii_base_for_version(dict_version)
        compat_note = (f" [backwards-compat: encoded with v{dict_version}, "
                       f"decoded via frozen snapshot]")
        print(f"[spec_dec] Backwards compatibility: using frozen v{dict_version} "
              f"token table ({len(id_to_token):,} tokens).")
    else:
        # Newer file than our dict — warn and try current dict
        import warnings
        warnings.warn(
            f"File was encoded with dict v{dict_version} but current dict is "
            f"v{D.DICT_VERSION}. Tokens added in v{dict_version} may fail to "
            f"decode. Upgrade Spectrum Algo to the latest version.",
            UserWarning,
        )
        id_to_token = D.SPEC_ID_TO_TOKEN
        ascii_base  = D.SPEC_ID_ASCII_BASE
        compat_note = f" [WARNING: encoded with newer v{dict_version}]"

    # Decompress body
    body = raw[HEADER_SIZE:]
    try:
        raw_stream = zlib.decompress(body)
    except zlib.error as e:
        raise SpecFormatError(f"zlib decompression failed: {e}")

    # Unpack uint32 LE stream (upgraded from uint16 in v7)
    count = len(raw_stream) // 4
    ids = list(struct.unpack(f"<{count}I", raw_stream[:count * 4]))

    # Decode to tokens (using version-appropriate ID table)
    tokens = ids_to_tokens(ids, id_to_token=id_to_token, ascii_base=ascii_base)

    # Reconstruct source — text files need the state-machine reconstructor
    # to interpret CTRL:* control tokens; all other languages just join.
    if meta["language_id"] in (LANGUAGE_TEXT, LANGUAGE_XML):
        source = reconstruct_text(tokens)
    else:
        source = "".join(tokens)

    # Truncate to original byte length
    source_bytes = source.encode("utf-8")
    if len(source_bytes) > orig_length:
        source = source_bytes[:orig_length].decode("utf-8", errors="replace")

    decoded_length = len(source.encode("utf-8"))

    # Verify checksum
    actual_cksum = sum(source.encode("utf-8")) & 0xFFFF
    cksum_ok = (actual_cksum == expected_cksum)

    # Write output
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(source, encoding="utf-8")

    length_ok = (decoded_length == orig_length)
    fidelity  = "✓ perfect" if (length_ok and cksum_ok) else "✗ mismatch"

    print(f"[spec_dec] Decoded {len(tokens):,} tokens → {decoded_length:,} bytes  "
          f"checksum {'✓' if cksum_ok else '✗'}  [{fidelity}]{compat_note}")
    print(f"[spec_dec] Saved → {output_path.name}")

    return {
        "spec_path":       str(spec_path),
        "output_path":     str(output_path),
        "dict_version":    dict_version,
        "orig_length":     orig_length,
        "decoded_length":  decoded_length,
        "token_count":     len(tokens),
        "length_ok":       length_ok,
        "checksum_ok":     cksum_ok,
        "fidelity":        fidelity,
    }


# ─────────────────────────────────────────────────────────────────────────────
# CLI
# ─────────────────────────────────────────────────────────────────────────────

def main():
    import argparse
    parser = argparse.ArgumentParser(
        description="Spectrum Algo .spec Decoder — binary → source")
    parser.add_argument("spec_file", help="Path to .spec file")
    parser.add_argument("--out", default=None,
                        help="Output file path (default: spec_format/output/<stem>_decoded.py)")
    args = parser.parse_args()

    spec = Path(args.spec_file)
    if not spec.exists():
        print(f"Error: {spec} not found", file=sys.stderr)
        sys.exit(1)

    if args.out:
        out = Path(args.out)
    else:
        out_dir = Path(__file__).resolve().parent / "output"
        out = out_dir / (spec.stem + "_decoded.py")

    result = decode_file(str(spec), str(out))
    if not (result["length_ok"] and result["checksum_ok"]):
        sys.exit(1)


if __name__ == "__main__":
    main()

File: chrome-extension/spec-decoder.js
/**
 * Spectrum .spec Decoder — ES Module
 * Shared by background.js (service worker) and viewer.js.
 *
 * Supports:
 *   v6  —  uint16 LE tokens, 447-token dictionary (inline)
 *   v7-v10 — uint32 LE tokens, append-only token dictionary loaded from spec-tokens.json
 *
 * Exports:
 *   decodeSpec(arrayBuffer)  → { source, meta, checksumOk, tokenCount }
 *   inferContentType(pathname) → MIME type string
 *   inferHljsLang(pathname)    → highlight.js language name
 */

// ── v6 inline dictionary (447 tokens) ────────────────────────────────────────
const SPEC_TOKENS_V6 = ["if", "elif", "else", "for", "while", "break", "continue", "pass", "return", "yield", "def", "class", "lambda", "async", "await", "import", "from", "as", "try", "except", "finally", "raise", "assert", "True", "False", "None", "and", "or", "not", "in", "is", "global", "nonlocal", "del", "with", "print", "+", "-", "*", "/", "//", "%", "**", "==", "!=", "<", ">", "<=", ">=", "=", "+=", "-=", "*=", "/=", "&", "|", "^", "~", "<<", ">>", "(", ")", "[", "]", "{", "}", ":", ",", ".", ";", "@", "#", "->", "...", "\"", "'", "\"\"\"", "'''", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", " ", "\t", "\n", "\r", "abs", "aiter", "all", "anext", "any", "ascii", "bin", "breakpoint", "callable", "chr", "compile", "copyright", "credits", "delattr", "dir", "divmod", "enumerate", "eval", "exec", "exit", "filter", "format", "getattr", "globals", "hasattr", "hash", "help", "hex", "id", "input", "isinstance", "issubclass", "iter", "len", "license", "locals", "map", "max", "memoryview", "min", "next", "oct", "open", "ord", "pow", "property", "quit", "range", "repr", "reversed", "round", "setattr", "slice", "sorted", "staticmethod", "sum", "super", "vars", "zip", "bool", "bytearray", "bytes", "classmethod", "complex", "dict", "float", "frozenset", "int", "list", "object", "set", "str", "tuple", "type", "self", "cls", "args", "kwargs", "result", "value", "name", "data", "key", "obj", "func", "text", "node", "msg", "path", "url", "mode", "error", "index", "size", "count", "buf", "tmp", "flag", "Exception", "ValueError", "TypeError", "KeyError", "IndexError", "AttributeError", "RuntimeError", "StopIteration", "IOError", "OSError", "ImportError", "NameError", "NotImplementedError", "OverflowError", "ZeroDivisionError", "FileNotFoundError", "PermissionError", "TimeoutError", "ConnectionError", "RecursionError", "MemoryError", "SyntaxError", "UnicodeError", "ArithmeticError", "append", "extend", "insert", "remove", "pop", "clear", "sort", "reverse", "get", "update", "keys", "values", "items", "join", "split", "strip", "replace", "startswith", "endswith", "encode", "decode", "read", "write", "close", "seek", "flush", "copy", "lower", "upper", "match", "search", "findall", "os", "sys", "re", "io", "math", "json", "abc", "ast", "time", "functools", "itertools", "collections", "pathlib", "typing", "logging", "threading", "subprocess", "datetime", "unittest", "html", "head", "body", "div", "span", "ul", "ol", "li", "table", "tr", "td", "th", "form", "button", "script", "style", "link", "meta", "title", "header", "footer", "nav", "section", "article", "main", "h1", "h2", "h3", "h4", "h5", "h6", "br", "hr", "strong", "em", "code", "pre", "img", "iframe", "canvas", "video", "audio", "href", "src", "alt", "placeholder", "rel", "disabled", "checked", "defer", "charset", "onclick", "onchange", "onsubmit", "onload", "tabindex", "download", "hidden", "multiple", "autocomplete", "colspan", "rowspan", "var", "let", "const", "function", "this", "new", "typeof", "instanceof", "switch", "case", "void", "debugger", "do", "of", "export", "null", "undefined", "NaN", "Infinity", "===", "!==", "=>", "++", "--", "?.", "??", "console", "document", "window", "JSON", "Object", "Array", "Math", "Number", "String", "Boolean", "Promise", "fetch", "setTimeout", "setInterval", "clearTimeout", "clearInterval", "addEventListener", "querySelector", "getElementById", "callback", "resolve", "reject", "prototype", "module", "@media", "@import", "@keyframes", "@font-face", "@supports", "@charset", "@layer", "@container", "@page", "@namespace", "display", "position", "top", "right", "bottom", "left", "width", "height", "max-width", "min-width", "max-height", "min-height", "overflow", "z-index", "margin", "margin-top", "margin-right", "margin-bottom", "margin-left", "padding", "padding-top", "padding-right", "padding-bottom", "padding-left", "border", "border-radius", "color", "font-size", "font-weight", "font-family", "line-height", "text-align", "text-decoration", "background", "background-color", "background-image", "opacity", "transform", "transition", "cursor", "none", "auto", "block", "inline", "inline-block", "flex", "grid", "absolute", "relative", "fixed", "sticky", "bold", "normal", "inherit", "initial", "unset", "center", "visible", "pointer", "solid"];
const SPEC_ID_ASCII_BASE_V6 = 447;
const SPEC_ID_RLE_V6        = 0xFFFD;
const SPEC_ID_UNICODE_V6    = 0xFFFE;

// ── v7 dictionary (lazy-loaded from spec-tokens.json) ─────────────────────────
// SPEC_ID_RLE_V7 = 0xFFFFFFFD, SPEC_ID_UNICODE_V7 = 0xFFFFFFFE
// uint32 LE tokens; Unicode escape followed by ONE uint32 code point
const SPEC_ID_RLE_V7     = 0xFFFFFFFD; // 4294967293
const SPEC_ID_UNICODE_V7 = 0xFFFFFFFE; // 4294967294

const CURRENT_DICT_VERSION = 10;
const ASCII_BASE_BY_DICT_VERSION = {
  7: 234702,
  8: 234830,
  9: 234893,
  10: 234957,
};

let _v7Tokens = null;  // Array<string>, current append-only token table

async function loadV7Tokens() {
  if (_v7Tokens !== null) return;
  const url = (typeof chrome !== 'undefined' && chrome.runtime?.getURL)
    ? chrome.runtime.getURL('spec-tokens.json')
    : new URL('./spec-tokens.json', import.meta.url).href;
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`Failed to load token dictionary (HTTP ${resp.status})`);
  _v7Tokens = await resp.json();
  if (_v7Tokens.length < ASCII_BASE_BY_DICT_VERSION[CURRENT_DICT_VERSION]) {
    throw new Error(`Token dictionary is stale (${_v7Tokens.length} entries).`);
  }
}

// ── Content-type inference from double extension ──────────────────────────────
const EXT_MIME = {
  '.html.spec':  'text/html',
  '.htm.spec':   'text/html',
  '.css.spec':   'text/css',
  '.js.spec':    'application/javascript',
  '.mjs.spec':   'application/javascript',
  '.cjs.spec':   'application/javascript',
  '.py.spec':    'text/plain',
  '.json.spec':  'application/json',
  '.ts.spec':    'application/javascript',
  '.tsx.spec':   'application/javascript',
  '.sql.spec':   'text/plain',
  '.rs.spec':    'text/plain',
  '.php.spec':   'text/plain',
  '.phtml.spec': 'text/plain',
  '.xml.spec':   'application/xml',
  '.md.spec':    'text/plain',
  '.txt.spec':   'text/plain',
};

export function inferContentType(pathname) {
  const p = pathname.toLowerCase().replace(/\?.*$/, '');
  for (const [ext, mime] of Object.entries(EXT_MIME)) {
    if (p.endsWith(ext)) return mime;
  }
  return 'text/plain';
}

export function inferHljsLang(pathname) {
  const p = pathname.toLowerCase();
  if (p.endsWith('.html.spec') || p.endsWith('.htm.spec')) return 'html';
  if (p.endsWith('.css.spec'))  return 'css';
  if (p.endsWith('.js.spec') || p.endsWith('.mjs.spec')) return 'javascript';
  if (p.endsWith('.ts.spec') || p.endsWith('.tsx.spec')) return 'typescript';
  if (p.endsWith('.py.spec'))   return 'python';
  if (p.endsWith('.xml.spec'))  return 'xml';
  if (p.endsWith('.php.spec') || p.endsWith('.phtml.spec')) return 'php';
  if (p.endsWith('.sql.spec'))  return 'sql';
  if (p.endsWith('.rs.spec'))   return 'rust';
  return 'plaintext';
}

// ── Header parsing ────────────────────────────────────────────────────────────
const MAGIC       = [0x53, 0x50, 0x45, 0x43]; // 'SPEC'
const HEADER_SIZE = 16;
const FLAG_RLE    = 0x01;

function readU16BE(buf, off) { return (buf[off] << 8) | buf[off + 1]; }
function readU32BE(buf, off) {
  return (buf[off] * 0x1000000) + (buf[off+1] << 16) + (buf[off+2] << 8) + buf[off+3];
}

function parseHeader(buf) {
  if (buf.length < HEADER_SIZE) throw new Error(`File too short (${buf.length} B).`);
  for (let i = 0; i < 4; i++) {
    if (buf[i] !== MAGIC[i]) throw new Error(`Bad magic bytes — not a .spec file.`);
  }
  const flags = readU16BE(buf, 6);
  return {
    dictVersion: readU16BE(buf, 4),
    flags,
    origLength:  readU32BE(buf, 8),
    languageId:  readU16BE(buf, 12),
    checksum:    readU16BE(buf, 14),
    rleEnabled:  !!(flags & FLAG_RLE),
  };
}

// ── zlib decompress (browser-native DecompressionStream) ─────────────────────
async function zlibDecompress(bytes) {
  const ds     = new DecompressionStream('deflate');
  const writer = ds.writable.getWriter();
  const reader = ds.readable.getReader();
  writer.write(bytes).catch(() => {});
  writer.close().catch(() => {});
  const chunks = [];
  for (;;) {
    const { value, done } = await reader.read();
    if (done) break;
    chunks.push(value);
  }
  const total = chunks.reduce((s, c) => s + c.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) { out.set(c, off); off += c.length; }
  return out;
}

// ── v6 token stream → source text (uint16 LE) ────────────────────────────────
function idsToSourceV6(ids) {
  const base   = SPEC_ID_ASCII_BASE_V6;
  const tokens = SPEC_TOKENS_V6;
  const parts  = [];
  let lastTok  = null;
  let i        = 0;

  while (i < ids.length) {
    const v = ids[i];

    if (v === SPEC_ID_RLE_V6) {
      const n = ids[i + 1];
      if (lastTok !== null) for (let k = 0; k < n; k++) parts.push(lastTok);
      i += 2; continue;
    }

    if (v === SPEC_ID_UNICODE_V6) {
      if (i + 2 >= ids.length) throw new Error(`Unicode escape at pos ${i} is truncated.`);
      // Use multiplication to avoid signed 32-bit overflow from << 16
      const cp  = ids[i + 1] * 65536 + ids[i + 2];
      const tok = (cp <= 0x10FFFF) ? String.fromCodePoint(cp) : '\uFFFD';
      parts.push(tok); lastTok = tok; i += 3; continue;
    }

    if (v >= base && v < base + 128) {
      const tok = String.fromCharCode(v - base);
      parts.push(tok); lastTok = tok; i++; continue;
    }

    const tok = tokens[v];
    if (tok === undefined) throw new Error(`Unknown v6 token ID ${v} at pos ${i}`);
    parts.push(tok); lastTok = tok; i++;
  }
  return parts.join('');
}

// ── v7 token stream → parts array (uint32 LE) ────────────────────────────────
// Returns an Array<string> of raw token strings (control tokens included).
// Callers decide whether to join() directly or pass through reconstructText().
function idsToPartsV7(ids, dictVersion) {
  const base   = ASCII_BASE_BY_DICT_VERSION[dictVersion];
  const tokens = _v7Tokens;
  const parts  = [];
  let lastTok  = null;
  let i        = 0;

  while (i < ids.length) {
    const v = ids[i];

    // RLE: 0xFFFFFFFD followed by uint32 repeat count
    if (v === SPEC_ID_RLE_V7) {
      const n = ids[i + 1];
      if (lastTok !== null) for (let k = 0; k < n; k++) parts.push(lastTok);
      i += 2; continue;
    }

    // Unicode fallback: 0xFFFFFFFE followed by uint32 code point
    if (v === SPEC_ID_UNICODE_V7) {
      if (i + 1 >= ids.length) throw new Error(`Unicode escape at pos ${i} is truncated.`);
      const cp  = ids[i + 1];
      const tok = (cp <= 0x10FFFF) ? String.fromCodePoint(cp) : '\uFFFD';
      parts.push(tok); lastTok = tok; i += 2; continue;
    }

    // ASCII fallback: base + char code (0–127)
    if (v >= base && v < base + 128) {
      const tok = String.fromCharCode(v - base);
      parts.push(tok); lastTok = tok; i++; continue;
    }

    // Dictionary token
    const tok = tokens[v];
    if (tok === undefined || v >= base) {
      throw new Error(`Unknown v${dictVersion} token ID ${v} at pos ${i} (ascii base ${base})`);
    }
    parts.push(tok); lastTok = tok; i++;
  }
  return parts;
}

// ── Plain-text reconstruction — mirrors text_tokenizer.reconstruct_text() ────
// Interprets CTRL:* control tokens and reassembles the original cased text.
function _applyCap(word, capMode) {
  if (!word || !capMode) return word;
  if (capMode === 'first') return word[0].toUpperCase() + word.slice(1);
  if (capMode === 'all')   return word.toUpperCase();
  return word;
}

function reconstructText(parts) {
  const result   = [];
  let capMode    = null;   // null | 'first' | 'all'
  let spelling   = [];
  let inSpelled  = false;

  for (const tok of parts) {
    // ── Control tokens ─────────────────────────────────────────────────────
    if (tok === 'CTRL:CAP_FIRST')  { capMode = 'first'; continue; }
    if (tok === 'CTRL:CAP_ALL')    { capMode = 'all';   continue; }
    if (tok === 'CTRL:NUM_SEP')    { continue; }   // boundary only; no output

    if (tok === 'CTRL:BEGIN_WORD') {
      inSpelled = true;
      spelling  = [];
      continue;
    }
    if (tok === 'CTRL:END_WORD') {
      result.push(_applyCap(spelling.join(''), capMode));
      capMode   = null;
      inSpelled = false;
      spelling  = [];
      continue;
    }

    // ── Inside a spelled-out word ───────────────────────────────────────────
    if (inSpelled) { spelling.push(tok); continue; }

    // ── All other tokens (words, whitespace, punctuation, digits) ──────────
    result.push(_applyCap(tok, capMode));
    capMode = null;
  }
  return result.join('');
}

// ── Top-level decode ──────────────────────────────────────────────────────────
export async function decodeSpec(arrayBuffer) {
  const buf  = new Uint8Array(arrayBuffer);
  const meta = parseHeader(buf);

  const dictVersion = meta.dictVersion;
  if (dictVersion !== 6 && !(dictVersion in ASCII_BASE_BY_DICT_VERSION)) {
    throw new Error(`Unsupported dict version ${dictVersion}. Supported versions: v6-v${CURRENT_DICT_VERSION}.`);
  }

  // Load append-only v7+ token list if needed
  if (dictVersion !== 6) await loadV7Tokens();

  const rawStream = await zlibDecompress(buf.slice(HEADER_SIZE));

  let ids, tokenCount;
  if (dictVersion === 6) {
    // uint16 LE
    tokenCount = Math.floor(rawStream.length / 2);
    const dv   = new DataView(rawStream.buffer, rawStream.byteOffset, rawStream.byteLength);
    ids = new Array(tokenCount);
    for (let i = 0; i < tokenCount; i++) ids[i] = dv.getUint16(i * 2, true);
  } else {
    // uint32 LE (v7+)
    tokenCount = Math.floor(rawStream.length / 4);
    const dv   = new DataView(rawStream.buffer, rawStream.byteOffset, rawStream.byteLength);
    ids = new Array(tokenCount);
    for (let i = 0; i < tokenCount; i++) ids[i] = dv.getUint32(i * 4, true);
  }

  let source;
  if (dictVersion === 6) {
    source = idsToSourceV6(ids);
  } else {
    const parts = idsToPartsV7(ids, dictVersion);
    // Language 4 = text, 9 = XML/Wiki. Both use text control tokens.
    source = (meta.languageId === 4 || meta.languageId === 9)
      ? reconstructText(parts)
      : parts.join('');
  }

  // Truncate to original byte length
  const enc      = new TextEncoder();
  const srcBytes = enc.encode(source);
  if (srcBytes.length > meta.origLength) {
    source = new TextDecoder('utf-8', { fatal: false }).decode(
      srcBytes.slice(0, meta.origLength)
    );
  }

  // Verify checksum
  const actualBytes = enc.encode(source);
  let sum = 0;
  for (const b of actualBytes) sum = (sum + b) & 0xFFFF;

  return { source, meta, checksumOk: sum === meta.checksum, tokenCount };
}
